pkg/controller/deployment/util/deployment_util.go

```diff
@@ -701,6 +820,66 @@ func IsRollingUpdate(deployment *extensions.Deployment) bool {
 	return deployment.Spec.Strategy.Type == extensions.RollingUpdateDeploymentStrategyType
 }
 
+// IsDeploymentComplete considers a deployment to be complete once its desired replicas equals its
+// updatedReplicas and it doesn't violate minimum availability.
+func IsDeploymentComplete(deployment *extensions.Deployment, newStatus *extensions.DeploymentStatus) bool {
+	return newStatus.UpdatedReplicas == deployment.Spec.Replicas &&
+		newStatus.AvailableReplicas >= deployment.Spec.Replicas-MaxUnavailable(*deployment)
```

Comment:
> The first thing missing here is the check deployment.Status.ObservedGeneration >= deployment.Generation

This check doesn't make sense in the deployment controller because the deployment controller is setting ObservedGeneration. So you can think of it as being implicitly true at this point.

> Also when a deployment is complete, shouldnt the available replicas exactly equal the desired (Spec.replicas) ? The maxUnavailable is only taken into account while doing the rolling update. but aftet the rolling update has finished, the replicas should be exactly equal. Am i amissing something ?

maxUnavailable is part of the strategy you want to deploy your pods. If you allow partial availability then we only need to guarantee partial availability.